Reinstalled React so it's decoupled from server

speck14 opened the pull request:
Decoupled React front-end from Express backend.

speck14 committed:
fixed vulnerabilities

speck14 committed:
started eliminating callbacks in server using test file asyncservers.js

speck14 committed:
started working on removing callbacks

speck14 committed:
Improved refactoring of SQL for DRY-ness

speck14 committed:
refactored to remove callbacks

speck14 committed:
used force resolutions to fix json-schema vulnerability

speck14 committed:
reinstalled React so it's decoupled from server

speck14 committed:
added concurrently npm package to make react talk to express

speck14 committed:
wrapped fetch in React.useEffect

speck14 commented:
@sdkester
Since I had a couple of unsuccessful attempts to incorporate React, I gave the decoupled implementation of React in the app its own branch. From what I read, the other option was rolling back and losing the other React-implementation commits (which I wanted to be able to roll back to if this method of implementation didn't work, and to look back at as a reference).
But also now I have merging conflicts, which I haven't dealt with before. I have a feeling making a new branch wasn't the best choice. At some point I'll need some guidance on how best to address these merge conflicts, but for now I'm going to keep pushing to this branch.

speck14 committed:
display movie names as list

speck14 committed:
refactored MovieList class into function using React hooks

speck14 committed:
started rendering movie item view & organized file names

speck14 committed:
Get and display movie genres in movie item view

speck14 committed:
fixed infinite rerender issue using useCallback

speck14 committed:
Delete movies.sqlite

speck14 committed:
Render movie item view

speck14 committed:
Merge branch 'react-assignment' of https://github.com/speck14/movie-app into react-assignment

speck14 commented:
@sdkester - I'm a bit stuck and need a bit of guidance when you have the time.
Here's a summary of what I'm working on, and where my issue is:

## Within the movie item view:
Steps:
1)useEffect is called on render (re-render triggered when match.params.pk changes)
2) useEffect triggers a call to getMovieInfo, which does a get request to /genres/pk

-  Get movie name and genre_fks from JSON response, set movieItem state to that data via setMovieItem

3)getMovieInfo then calls getCurrentMovieGenres

- Makes a get request to /genres to get all genre information

- **_Sets allGenres to JSON response via setAllGenres_**

    - **This is where the app isn't working.** The request is successful, the JSON response contains genre data, but the setAllGenres call isn't setting allGenres to the JSON data

    - The setAllGenres works if I'm already on the movieItemView page and restart the app (...most of the time)

    - It doesn't work when navigating from the movieView page via the linked movie titles, and it doesn't work if the the page is refreshed

    - Calls are happening in the correct order, so I don't *think* it's an async/call stack order problem

4) getAllGenres then calls getCurrentMovieGenres

-  This uses the genre_fks from movieItem, compares it to pks for all genres, and adds matching pk/fks to movieGenres via setGenres
5)Display movie name and a list of genres for current movie

speck14 committed:
code clean-up

sdkester commented:
@speck14 I tried to take a look at this, but all of the react components got [taken out of your package.json file](https://github.com/speck14/movie-app/blob/react-assignment/package.json), so this branch doesn't run. I suspect the merge at 401036b, but I'm not sure. Please ping me Wednesday once you restore the packages and we can take a look at it together.

speck14 commented:
@sdkester There's a [second package.json](https://github.com/speck14/movie-app/blob/react-assignment/client/package.json) located in the client directory that has the React components. The [one in the root directory](https://github.com/speck14/movie-app/blob/react-assignment/package.json) contains the SQLite and Express components.
I found that having separate package.json files for front end and back end seemed to be the way most tutorials and articles handled React.

speck14 commented:
@sdkester 
I've spent some more time reading into React hooks/state (with the intention of moving on to creating the "add movie" form for now). But, that led me to a potential path for my issue.

I think my problem lies with how I'm initializing the movie item's state as an object; from what I noticed this afternoon, my getCurrentMovieGenres function is "looking" at the empty genre_fks array from that initialized state, and not the state I "get" from my fetch. Would setting my state with the useReducer hook (instead of useState) potentially fix that?

speck14 committed:
started 'addMovie' form, refactored movieitem.js

speck14 committed:
combined React movie list and item view into one functional component

speck14 commented:
@sdkester 
I was able to combine the movie list and individual movie views into one file/React component, and now the list view is displaying the genres!
Next up I'm going to figure out the issues with initially starting the app locally, and update the "ReadMe" accordingly.

speck14 committed:
fix local install problems

speck14 committed:
added postinstall script to package.json

speck14 committed:
updated readme

speck14 commented:
@sdkester 
The problem with running the app on fresh installations goes all the way back to when I refactored to removed callbacks.
I didn't realize at the time, but my refactoring passed the test because I had an existing movies.sqlite that was passing the tests. I added that to the .gitignore some time back, but movies.sqlite was still in my local repo. So, my local repo passed the tests because I had a  previous database that passed. Likewise, since movies.sqlite is no longer being cloned when first installing, new repos no longer pass the test (because my refactoring to eliminate callbacks was not actually successful).

I attempted to figure out where my refactoring went wrong, but was unable to successfully use async.each to set up the DB. In the server.js file, at line 126 you can see where I started to go back and refactor. The database successfully sets up if I individually pass in the SQL for the table, but not if I pass it in as an async.each. However, inserting the genres into the appropriate table works just fine when passed as async.each.
I'm stumped as to why my async calls aren't working for table creation- but I included the async functions I tried as a comment (starting line 143 in server.js), if you could take a look and give me some guidance of what I'm missing here.

speck14 committed:
capture user input, render to browser in add movie form

sdkester commented:
@speck14 You forgot to add callbacks to the iterator function of [each method.](https://caolan.github.io/async/v3/docs.html#each) 

They should follow this pattern:

```javascript
console.log(`iterate some things`)
async.each(array, (item, callback) => {
  console.log(`this is ${item}`)
  calllback()
}, (err)=>{
  console.log(`done with all items`)
})
```

Anytime an asyncJS function isn't working as expected, put console.log statements at each layer and see that you 1) hit all of them and 2) hit them in the expected order.

Once you have added callbacks and some `console.logs()`, I expect your terminal to look like this:

![image](https://user-images.githubusercontent.com/481473/148135710-cff2178a-42d2-46ca-aac4-cf778ffbe6dc.png)

If you need to step through your server.js code, you can add this to your package.json and then start the server with `npm run debug-server`:

```javascript
"debug-server": "node --inspect-brk=5858 server.js"
```

Then visit the url `chrome://inspect/#devices` in the Chrome web browser and click the _inspect_ link below server.js

![image](https://user-images.githubusercontent.com/481473/148135902-45831016-9c23-41df-b7a2-415a6cc9b5c7.png)

This will give you the opportunity to add break points and step through the code.

speck14 committed:
fixed async.each call

speck14 committed:
refactored addMovie component to eliminate react routing

speck14 committed:
display genres as checkbox input

speck14 committed:
added checkboxes to add movie form

speck14 committed:
add movie functionality

speck14 committed:
fixed formatting

speck14 committed:
refactored reusable components, started 'edit movie' form

speck14 committed:
initially check boxes in 'edit form' for current genres. Delete functionality

speck14 committed:
added comments

speck14 committed:
implemented full CRUD functionality

speck14 commented:
@sdkester As of this afternoon, the app has full CRUD functionality! (And therefore is doing everything the assignment asked).

speck14 committed:
Refactored 'movie list' component

speck14 committed:
Update readme.md

sdkester commented on package.json:
```diff
@@ -8,18 +8,30 @@
+    "preinstall": "npx npm-force-resolutions",
+    "postinstall": "cd client && npm install",
+    "client": "npm start --prefix client",
+    "dev": "concurrently \"nodemon --inspect=5858 server.js\" \"npm run client\"",
```
@speck14 It'd be important to note in the README that you are relying on nodemon being installed globally.  Just a small block about installing it like `npm install -g nodemon` if you don't already have it.

sdkester commented:
@speck14 It's working great for me, I'm so excited to see it all come together. Nice work! Your problem solving skills shined brightly here. 

I've left a message with the team inviting them to do a code review for your benefit and ours. This could be a good basis for an updated interview exercise we might use in the future.

speck14 committed:
Update readme.md

speck14 commented:
@MGVining
Thanks so much for agreeing to review this!

MGVining commented on client/src/button.js:
```diff
@@ -0,0 +1,7 @@
+
+function Button({ clickHandler, text }) {
+  return <button onClick={clickHandler}>{text}</button>;
+}
```
In production code, I would not abstract a component that's this small.

Until there is some repeated static stuff on each button, it isn't worth pulling it out.
As an example, if every button needed identical styling and auth code, that could live here.

servant-42 replied:
> In production code, I would not abstract a component that's this small.

@speck14: :100:  I can't tell you how many times I've had to fight through a bunch of unnecessary layers when debugging/maintaining code. As the [Zen of Python](https://www.python.org/dev/peps/pep-0020/) says "Flat is better than nested", or as [Isaac Schleuter says](https://blog.izs.me/2008/01/going-fast-frankenstein-and-refactoring/) "There should be as few layers as possible, and no fewer".

@MGVining's rule is good: "Until there is some repeated stuff, it's not worth pulling it out". DRY is a good way to know when something needs to be separated out. Casey Muratory calls this [Semantic Compression](https://caseymuratori.com/blog_0015) -- fixing DRY violations as a way to know where the layers & code-reuse needs to be, and avoiding it everywhere it doesn't need to be.

MGVining commented on client/src/checkbox.js:
```diff
@@ -0,0 +1,22 @@
+          name={name}
+          value={id}
+          checked={isChecked}
+          onChange={() => eventHandler(index)}
```
`index` and `eventHandler` are only passed down to this component to be combined and passed back up.

Combining them feels like it should be done in the parent rather than in the Checkbox, that way we prevent passing around the `index` data when it isn't needed down here.

speck14 replied:
I tried wrapping each checkbox in a div and putting the handling logic there:
```
{allGenres.map((genre, index) => (
            <div className="clickHandler" onClick={handleCheckChange(index)}>
            <Checkbox
              key={genre.pk}
              name={genre.name}
              id={genre.pk}
              //index={index}
              isChecked={checkedState[index]}
              //eventHandler={handleCheckChange}
            />
            </div>
```
But it results in a "Maximum updated depth exceeded" error.
Where would be the best place to combine `index` and `eventHandler`? I'm having a hard time seeing it.

MGVining replied:
"Maximum updated depth exceeded" usually means infinite loop. In this case looks like an infinite render loop.
The new `onClick={handleCheckChange(index)}` has the parenthesis at the end, which invokes the function, which re-render the component, which invokes the function, which re-renders until you crash.

I would go for something like:
```
1 {allGenres.map((genre, index) => (
2    var changeHandler = () => {handleCheckChange(index)}
3    return <Checkbox
4        ....
5        eventHandler={changeHandler}
```

* You can add regular function lines like 2. to the arrow function on 1. as long as you denote a return on 3.
* Line 2 is similar to the handler you attached to a `div`, but doesn't invoke the handleCheckChange early, since it's wrapped in another arrow function.
* Line 5 We want to pass the handler to checkbox so it can manage when to fire the handler, but it no longer needs to know how to mix the handler with index.

speck14 replied:
Oh I see, that totally makes sense. Thank you so much.

MGVining commented on client/src/movieList.js:
```diff
@@ -0,0 +1,37 @@
+                  movieClick();
+                }}
+              >
+                <Movie name={movie.name} pk={movie.pk} />
```
Recommend completing or removing the abstraction here. I'm fond of grouping DOM elements into one location on the page.

Every `Movie` needs an `<li>` element, so you move that into the variable above.
Every `Movie` also needs a `<button>`, so you can move the whole button element up there too (adding props as needed).
That way this section reads:
`{allMovies.map((movie) => ( <Movie name={movie.name} otherPropsYouNeed={stuffAndThings} /> )};`
With all HTML grouped together in the variable.

Alternatively, remove the `Movie` abstraction and bring the `<li>` back down into the DOM.

MGVining commented on client/src/editMovie.js:
```diff
@@ -0,0 +1,139 @@
+    return initSelectedGenres;
+  }
+
+  async function updateMovie() {
```
In production, I would move a function like this up to app.js and pass a reference to it down through props or context.

editMovie doesn't need to know the details of how our fetch works, it just needs to be able to use it. By keeping all of the low level program details like the fetch URL up in app.js, there's only one place we need to look if we change it.

MGVining reviewed (commented):
Reviewed in the context of looking for react being used in react-y ways.

You're using functional react here, which Kaung tells me is the cooler way to play. I'm more familiar with class component, so I can't comment too much on your use of effects/state.
Data flow looks good, server interaction looks good. Some comments about when to abstract.
Overall nice work!

speck14 committed:
removed button component abstraction

speck14 committed:
Merge branch 'react-assignment' of https://github.com/speck14/movie-app into react-assignment

speck14 committed:
completed abstraction of 'Movie' from movieList

speck14 committed:
move 'updateMovie' function out of editMovie component

speck14 committed:
no longer passing index, event handler into Checkbox

kndshein commented:
Woohoo, great work @speck14 ! The app has come very far since the last time we chatted. Not sure what you did to fix it, but installing the project locally was flawless. I decided to clone the project and go through the code with the app open instead of going through them on GitHub, which allow me to view the app holistically and also _show_ you my review. I've divvied them up with commits [here](https://github.com/speck14/movie-app/tree/kaung-review) along with the steps provided below.

#### General feedback
- Use `let` and `const` instead of `var`. Don't tell PR I mentioned `const`, but from my experience, most folks use `const` to declare components in React. *shrug*
- When passing props, I like to keep the naming the same. This allows me to keep track of the names of the props since it'll be the same in both the parent and the children components. By giving additional names, not only are we abstracting the props, we also need to come up with more variables, adding complexity. This general rule of thumb, of course, has exceptions -- particularly when there's a difference in prop names from different components being passed down to is a generic component.
- The way you're using CSS reminds me of tailwindCSS, check it out! (I personally don't use it -- I like to push all the pixels manually, but that's neither here nor there)
- Destructuring props -- I'm a big fan. Well, kind of. I like things as concise as possible, but lately I've noticed that the prefix `props.` helps a lot differentiating component variables from passed props. On one hand, an argument can be made that components by nature should be compact, thus the confusion shouldn't be there. On the other, in bigger projects, I've noticed that some components can get big, and it's a pain to mentally remember which variables are props vs. component variables. Not an advise on what you should use, just food for thought!
- Variable naming
    - Be consistent with naming things. Nouns vs verbs in `handle...` and `...handler`. Keep it consistent. In React (and maybe JavaScript as a whole), it's more common to use verbs than nouns -- [Relevant reading, but not for React](https://steve-yegge.blogspot.com/2006/03/execution-in-kingdom-of-nouns.html)
    - PascalCase is considered best practice for naming components in React -- I see that you've done that w/ the components themselves, but they should be reflected in the file names as well.
- Not sure if this project is done, but make sure to remove commented out code or cruft. It's mighty ballsy coming from me considering I'm terrible at that, but still something to strive for
- Instead of `export default <ComponentName>` at the end of the file, you can simple move that in the function declaration. (i.e. `export default function <ComponentName>({...}))`). This, however, cannot be done w/ arrow functions. Why? Good question, I like that you're curious. [Relevant SO](https://stackoverflow.com/questions/34676984/cannot-export-const-arrow-function)

#### What I did step by step
Unfortunately, due to time constraint, I wasn't able to do a 100% deep clean. The few components I wasn't able to look closely were `Form.js`, `EditMovie.js` and `AddMovie.js`. Thanks for letting me review this! This refresher will be come in very handy for me when I have to start using React in IAC. :)

1. In `app.js` I refactored the if/else chain to the logical && operator -- this is best practice for conditional rendering. This does remove the logical series nature (i.e. if the conditionals are set up wrong, multiple things can be rendered), but there are so few conditions and this is much cleaner, I'd use this.
1. Replaced `div` tags in `app.js` wrapping each components into [React Fragments](https://reactjs.org/docs/fragments.html) lest we contract [divitis](https://en.wiktionary.org/wiki/divitis)
1. Passed `handleBackClick` into the components themselves that are using it. Also removed `preventDefault()` since it's not a form.
1. It's important to avoid `useRef` at all costs. So, I refactored the `useRef` to `useState`. Dealing with multiple `useState` in conjunction can be tricky, but something that's worth it to avoid `useRef`. As we're discussed prior, `useReducer` would be preferred here. Regardless, I removed useRef from `selectedMovie` and `selectedMovieGenres`. The async function wrapper is also removed.
1. Removed passing of `selectedMovie` and `setSelectedMovie` to `MovieList`. The former was from your code, and the latter was my addition -- both no longer necessary in `MovieList` since their usages could be consolidated in `app.js` in `handleMovieViewClick`. It's important to reduce passing props as much as possible -- I'm sure you're aware of it, but I want to take the time to drive the point home. *insert golf emoji* Not to drill into this further, but prop drilling is a real issue (i.e. passing props multiple layers without using said prop in the middle layers). `useContext`, a React Hook, solves this problem -- feel free to check it out when you're more comfortable.
1. Cleaned up unused variables and imports (i.e. handleAddMovie)
1. `MovieList.js` is housing two components. Usually a bad practice. Create components in its own file. In this case though, `Movie` is so small, I de-abstracted it.
1. Step 7, but for `Genres` in `MovieItem.js`
1. I see that your code has split `selectedMovie.name` and `selectedMovie.pk` into `selectedMovieTitle` and `selectedMoviePK`. This creates two different variables from something that's already coupled, which adds complexity. 1. needing to pass two props instead of one. 2. needing to keep track of two variables instead of one object.
1. Here's a live example of the second point in "General Feedback."
    - In `app.js`, there's `setViewMovie` from a `useState`.
    - The `setViewMovie` is then wrapped in `movieViewWrapper`
    - The wrapper is then passed as `movieViewSetter` into `MovieItem.js`
    - So, now the code has `movieViewSetter` in `MovieItem.js` -- but in order to know what it is, I have to follow the thread all the way back to `app.js`
    - Simple pass `setViewMovie` all the way down. Not only does this avoid two variables, but also allow other devs to find what it does right away
    - Same thing applies for renaming `updateMovie` to `movieUpdater` and `allMovieGenres` to `allGenres`.
1. Shutting down the concert of wrappers in `AddMovie.js` and `EditMovie.js`
    - I did not rename `titleChangeHandler` to `setMovieTitle` (though `titleChangeHandler` should be renamed to `handleTitleChange`, see "variable naming" in "General Feedback") because I noticed that you're reusing the `Form.js` component for `AddMovie.js` and `EditMovie.js` -- bravo!
    - Additionally, notice the ternary operator change in `handleClearClick`
1. In `app.js`, both async/await and then/catch. Use try/catch with async/await instead. [A Comparison Of async/await Versus then/catch](https://www.smashingmagazine.com/2020/11/comparison-async-await-versus-then-catch/)
1. Good attempt to fix MV's [comment](https://github.com/speck14/movie-app/pull/2#discussion_r797941840). However, you're [still passing the index](https://github.com/speck14/movie-app/pull/2#discussion_r797941840), making the fix void. Not to undercut MV here, but you can further consolidate the function MV provided.
1. Great touch on setting up the logic to clear the movie title if the form was accessed through "Add Movie" but retaining the title if accessed via "Edit Movie" -- great developers work hand in hand w/ UX, something I personally look up to and strive for.
    - Not sure why the `initTitle` in `Form.js` is an array. I reckon it's a typo.

#### Future Research/Work
Here's a list of things you can now learn/check out! I'm still learning some of them atm. Though not all of them are useful, I like to somewhat familiar with them.
- Great start of owning `useState` and `useEffect`. React offers [so much more](https://reactjs.org/docs/hooks-reference.html)!
    1. useReducer
    1. useContext
    1. useCallback
    1. useMemo
    1. Redux (not a React hook, but a library) -- I personally use a combination of `useReducer` and `useContext` and ignore Redux. But it's still good to know its ugly existence.
- React adjacent frameworks!
    1. Gatsby
    1. Next.js
- I know this was a React project and not a styling challenge, but check these out:
    1. TailwindCSS (again, not a big fan myself, but I can see why people like it)
    1. Module SCSS

Great work all in all! Especially doing a CRUD project for your very first React is a big challenge. Heck, even owning controlled forms in your first project on your own is a big one. I know I wrote a lot, but don't let that take away your accomplishment. Feel free to lmk if you have clarifying questions or particular topics that I didn't get on and you wish I did. Good luck!

speck14 commented:
@kndshein 
Thank you so much for taking the time to do such a detailed review. It's super valuable! I especially appreciate the branch with corresponding changes, it makes it much easier for me to see what you're suggesting.
As I go through everything in more detail, I'll let you know if I have any questions. In the meantime, I wanted to make sure I thanked your for the effort you clearly put into making this better!

prust commented:
> Don't tell PR I mentioned const

![always watching](https://media.giphy.com/media/KjdpBT4j79NJe/giphy.gif)

> from my experience, most folks use const to declare components in React

@kndshein @speck14: While I don't love `const` (at least, not yet), I'm also not strongly against it (`const` in C++ is another matter) and I agree that it is popular/idiomatic with React.

> Thank you so much for taking the time to do such a detailed review. It's super valuable! I especially appreciate the branch with corresponding changes

Ditto. Nice work, Kaung!
